File: CP264/cp264_a08_cabr9750/avl.c
/*
 -------------------------------------
 File:    avl.c
 Project: cabr9750_a07
 file description
 avl
 -------------------------------------
 Author:  Jacob Cabral
 ID:      190689750
 Email:   cabr9750@mylaurier.ca
 Version
 -------------------------------------
 */
#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include "avl.h"

// A utility functions
int max(int a, int b) {
	return (a > b) ? a : b;
}

int height(TNODE *np) {
	int height;
	height = 0;

	if (np) {
		height = np->height;
	}

	return height;
}

int balance_factor(TNODE *np) {
	int balance_factor;
	if (np) {
		balance_factor = np->left->height - np->right->height;
	} else if (np->left && !np->right) {
		balance_factor = np->left->height;
	} else if (!np->left && np->right) {
		balance_factor = 0 - np->right->height;
	} else {
		balance_factor = -1;
	}

	return balance_factor;
}

int is_avl(TNODE *root) {
	int result = 1;
	//check if there is no root
	if (!root) {
		result = 1;
	}
	//checking if the height of the root is less than the height of both left and right sub-trees
	if (height(root) < _max(height(root->left), height(root->right)) + 1) {
		result = 0;
	}
	if (abs(balance_factor(root)) > 1) {
		result = 0;
	}
	if (root->left && strcmp(root->left->data.name, root->data.name) > 0
			|| root->right
					&& strcmp(root->right->data.name, root->data.name) < 0) {
		result = 0;
	}
	if (!is_avl(root->left) || !is_avl(root->right)) {
		result = 0;
	}
	return result;
}

TNODE* rotate_right(TNODE *y) {
	TNODE *temp = y->left;
	TNODE *temp2 = temp->right;

	temp->right = y;
	y->left = temp2;

	y->height = 1 + _max(height(y->left), height(y->right));
	temp->height = 1 + _max(height(temp->left), height(temp->right));

	return temp;
}

TNODE* rotate_left(TNODE *x) {
	TNODE *temp = x->right;
	TNODE *temp2 = temp->left;

	temp->left = x;
	x->right = temp2;

	x->height = 1 + _max(height(x->left), height(x->right));
	temp->height = 1 + _max(height(temp->left), height(temp->right));

	return temp;
}

void insert(TNODE **rootp, char *name, float score) {
	TNODE *np = (TNODE*) malloc(sizeof(TNODE));
	if (np == NULL)
		return;
	strcpy(np->data.name, name);
	np->data.score = score;
	np->height = 1;
	np->left = NULL;
	np->right = NULL;

// 1. Perform the normal BST insertion
	if (*rootp == NULL) {
		*rootp = np;
		return;
	}

	TNODE *root = *rootp;
	if (strcmp(name, root->data.name) < 0)
		insert(&root->left, name, score);
	else if (strcmp(name, root->data.name) > 0)
		insert(&root->right, name, score);
	else
		return;

// 2. update height of this root node

// add your code here

// 3. Get the balance factor of this ancestor node to check whether this node became unbalanced

// add your code here

// 4. re-balance if not balanced

// add your code here

}

void delete(TNODE **rootp, char *name) {
	TNODE *root = *rootp;
	TNODE *np;

	if (root == NULL)
		return;

	if (strcmp(name, root->data.name) == 0) {
		if (root->left == NULL && root->right == NULL) {
			free(root);
			*rootp = NULL;
		} else if (root->left != NULL && root->right == NULL) {
			np = root->left;
			free(root);
			*rootp = np;
		} else if (root->left == NULL && root->right != NULL) {
			np = root->right;
			free(root);
			*rootp = np;
		} else if (root->left != NULL && root->right != NULL) {
			np = extract_smallest_node(&root->right);
			np->left = root->left;
			np->right = root->right;
			free(root);
			*rootp = np;
		}
	} else {
		if (strcmp(name, root->data.name) < 0) {
			delete(&root->left, name);
		} else {
			delete(&root->right, name);
		}
	}

// If the tree had only one node then return
	if (*rootp == NULL)
		return;

	root = *rootp;

// STEP 2: update the this root node's height

// add your code here

// STEP 3: get the balance factor of this root node

// add your code here

// STEP 4: re-balance if not balanced

// add your code here
}

// following functions are from bst.c of a7
TNODE* extract_smallest_node(TNODE **rootp) {
	TNODE *tnp = *rootp;
	TNODE *parent = NULL;
	if (tnp == NULL) {
		return NULL;
	} else {
		while (tnp->left) {
			parent = tnp;
			tnp = tnp->left;
		}
		if (parent == NULL)
			*rootp = tnp->right;
		else
			parent->left = tnp->right;
		tnp->left = NULL;
		tnp->right = NULL;
		return tnp;
	}
}

TNODE* search(TNODE *root, char *name) {
	TNODE *tp = root;
	while (tp) {
		if (strcmp(name, tp->data.name) == 0) {
			return tp;
		} else if (strcmp(name, tp->data.name) < 0)
			tp = tp->left;
		else
			tp = tp->right;
	}
	return NULL;
}

void clean_tree(TNODE **rootp) {
	if (*rootp) {
		TNODE *np = *rootp;
		if (np->left)
			clean_tree(&np->left);
		if (np->right)
			clean_tree(&np->right);
		free(np);
	}
	*rootp = NULL;
	;
}

void display_inorder(TNODE *root) {
	if (root) {
		if (root->left)
			display_inorder(root->left);
		printf("(%s %3.1f) ", root->data.name, root->data.score);
		if (root->right)
			display_inorder(root->right);
	}
}

void display_inorder_lines(TNODE *root) {
	if (root) {
		if (root->left)
			display_inorder_lines(root->left);
		printf("%-15s%.1f\n", root->data.name, root->data.score);
		if (root->right)
			display_inorder_lines(root->right);
	}
}

void display_tree(TNODE *root, int prelen) {
	if (root) {
		int i;
		for (i = 0; i < prelen; i++)
			printf("%c", ' ');
		printf("%s", "|___");
		printf("%s %.1f %d\n", root->data.name, root->data.score, root->height);
		display_tree(root->right, prelen + 4);
		display_tree(root->left, prelen + 4);
	}
}

File: cp213/cabr9750_a04/src/cp213/AVL.java
package cp213;

/**
 * Implements an AVL (Adelson-Velsky Landis) tree. Extends BST.
 *
 * @author your name here
 * @author David Brown
 * @version 2020-11-04
 */
public class AVL extends BST {

	/**
	 * Returns the balance value of node. If greater than 1, then left heavy, if
	 * less than -1, then right heavy. If in the range -1 to 1 inclusive, the node
	 * is balanced. Used to determine whether to rotate a node upon insertion.
	 *
	 * @param node The TreeNode to analyze for balance.
	 * @return A balance number.
	 */
	private int balance(final TreeNode node) {
		return this.nodeHeight(node.getLeft()) - this.nodeHeight(node.getRight());
	}

	/**
	 * Performs a left rotation around node.
	 *
	 * @param node The subtree to rotate.
	 * @return The new root of the subtree.
	 */
	private TreeNode rotateLeft(final TreeNode node) {
		// Rearrange the nodes.
		final TreeNode temp = node.getRight();
		node.setRight(temp.getLeft());
		temp.setLeft(node);
		node.updateHeight();
		temp.updateHeight();
		
		
		return temp;
	}

	/**
	 * Performs a right rotation around {@code node}.
	 *
	 * @param node The subtree to rotate.
	 * @return The new root of the subtree.
	 */
	private TreeNode rotateRight(final TreeNode node) {
		// Rearrange the nodes.
		final TreeNode temp = node.getLeft();
		node.setLeft(temp.getRight());
		temp.setRight(node);
		// Update the node heights.
		node.updateHeight();
		// Return new root.
		return temp;
	}

	/**
	 * Auxiliary method for {@code insert}. Inserts data into this AVL.
	 *
	 * @param node the current node (TreeNode)
	 * @param data Data to be inserted into the node
	 * @return The inserted node.
	 */
	@Override
	protected TreeNode insertAux(TreeNode node, final CountedCharacter data) {
		int balance = 0;

		if (node == null) {
			// Base case - add a new node containing the data.
			node = new TreeNode(data);	
			
			this.size++; 
			
			
		} else {
			// Compare the node data against the insert data.
			final int result = node.getValue().compareTo(data);

			if (result > 0) {
				// General case - check the left subtree.
				node.setLeft(this.insertAux(node.getLeft(), data));
				node.updateHeight();
				// As a left insertion, can only be unbalanced on the left.
				balance = this.balance(node);

				if (balance > 1 && this.balance(node.getLeft()) >= 0) {
					// Left Left Case - single rotation
					node = this.rotateRight(node);
				} else if (balance > 1 && this.balance(node.getLeft()) < 0) {
					// Left Right Case - double rotation
					node.setLeft(this.rotateLeft(node.getLeft()));
					node = this.rotateRight(node);
				}
			} else if (result < 0) {
				// General case - check the right subtree.
				node.setRight(this.insertAux(node.getRight(), data));
				node.updateHeight();
				// As a right insertion, can only be unbalanced on the right.
				balance = this.balance(node);

				if (balance < -1 && this.balance(node.getRight()) <= 0) {
					// Right Right Case - single rotation
					node = this.rotateLeft(node);
				} else if (balance < -1 && this.balance(node.getRight()) > 0) {
					// Right Left Case - double rotation
					node.setRight(this.rotateRight(node.getRight()));
					node = this.rotateLeft(node);
				}
			} else {
				// Base case - data is already in the tree, increment its count
				
				node.getValue().incrementCount();
				
			}
		}
		return node;
	}

	/**
	 * Auxiliary method for {@code valid}. Determines if a subtree based on node is
	 * a valid subtree. An AVL must meet the BST validation conditions, and
	 * additionally be balanced in all its subtrees - i.e. the difference in height
	 * between any two children must be no greater than 1.
	 *
	 * @param node The root of the subtree to test for validity.
	 * @return true if the subtree base on node is valid, false otherwise.
	 */
	@Override
	protected boolean isValidAux(final TreeNode node) {
		boolean valid = false;

		if (node == null) {
			// Base Case: bottom of tree
			valid = true;
		} else if (Math.max(this.nodeHeight(node.getLeft()), this.nodeHeight(node.getRight())) != node.getHeight()
				- 1) {
			// Base Case - height fault: current node height is not valid
			valid = false;
		} else if (node.getLeft() != null && node.getLeft().getValue().compareTo(node.getValue()) >= 0
				|| node.getRight() != null && node.getRight().getValue().compareTo(node.getValue()) <= 0) {
			// Base Case - data fault: current data not valid compared to
			// children's data
			valid = false;
		} else if (Math.abs(this.nodeHeight(node.getLeft()) - this.nodeHeight(node.getRight())) > 1) {
			// Base Case - AVL fault: child node heights are not correct
			valid = false;
		} else {
			valid = this.isValidAux(node.getLeft()) && this.isValidAux(node.getRight());
		}
		return valid;
	}

	/**
	 * Determines whether two AVLs are identical.
	 *
	 * @param target The AVL to compare this AVL against.
	 * @return true if this AVL and target contain nodes that match in position,
	 *         value, count, and height, false otherwise.
	 */
	public boolean equals(final AVL target) {
		return super.equals(target);
	}

}
